_posts/2018-07-05-modelling-page-view-events-as-a-graph/index.md

```diff
@@ -0,0 +1,490 @@
+
+    "required": ["schema", "data"],
+    "additionalProperties": false
+  }
+}
+```
+
+And for some events, such as `page_view`, there isn't a specific JSON schema. The [Elasticsearch enriched event schema][elasticsearch_enriched_event] probably comes closest, being a projection of the [canonical event model][canonical-event-model] (of which the `atomic.events` [table in Redshift][redshift-storage] is another projection). But that schema also applies to the other events in the canonical model.
+
+### Composable schemas
+
+For this project we'd like to explore a new approach to building schemas, based on individual schemas for each of the event's components, be they nodes or relationships. The schema for the event itself will then reference those building blocks through [JSON pointers][json-pointer]. This will allow us to reuse entities and relationships across all events they are part of. It will also make maintaining the schemas much easier. For example, if an entity such as a user is part of many events and we want to add a new property to the `User` node, we'll only have to make one change: in the `user` schema. (Currently our self-describing JSON schemas do not support JSON pointers, so this functionality will have to be [added][issue-370] if we decide to go with this approach.)
+
+To illustrate how this could work, let's look at a simplified `page_view` definition, consisting of two nodes -- `User` and `Page` -- connected by a `VISITED` relationship.
+
+We start by defining the schemas for each of the two nodes and the relationship:
+
+```JSON
+{
+  "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+  "description": "Schema for a User node",
+  "self": {
+    "vendor": "com.snowplowanalytics.snowplow",
+    "name": "user",
+    "format": "jsonschema",
+    "version": "1-0-0"
+  },
+
+  "type": "object",
+  "properties": {
+    "email": {
+      "type": "string"
+    }
+  },
+
+  "additionalProperties": false
+},
+
+{
+  "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+  "description": "Schema for a Page node",
+  "self": {
+    "vendor": "com.snowplowanalytics.snowplow",
+    "name": "page",
+    "format": "jsonschema",
+    "version": "1-0-0"
+  },
+
+  "type": "object",
+  "properties": {
+    "pageUrl": {
+      "type": "string"
+    }
+  },
+
+  "additionalProperties": false
+},
+
+{
+  "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+  "description": "Schema for a VISITED relationship",
+  "self": {
+    "vendor": "com.snowplowanalytics.snowplow",
+    "name": "visited",
+    "format": "jsonschema",
+    "version": "1-0-0"
+  },
+
+  "type": "object",
+  "properties": {
+    "derivedTstamp": {
+      "type": "string",
+      "format": "date-time"
+    },
+    "startNode": {
+      "$ref": "http://iglucentral.com/schemas/com.snowplowanalytics.graph/user/jsonschema/1-0-0#"
```

Comment:
HTTP URI (+json pointer) is a very low-level detail. If Iglu will ever implement `$ref`-analog - it will be a full `iglu:` URI. Also, do you think it is useful to have "JSON Pointer" part? For me it looks like we should be able to refer to the full self-describing schema as a complete entity, not to some "subschema". Would like to get your ideas also in the Iglu ticket.